Source/Core/Core/Lua/LuaEventCallbackClasses/LuaWheneverCallbackClass.cpp

```diff
@@ -0,0 +1,127 @@
+#include "Core/Lua/LuaEventCallbackClasses/LuaWheneverCallbackClass.h"
+
+#include <array>
+#include <condition_variable>
+#include <memory>
+#include <mutex>
+#include <random>
+#include <thread>
+#include <vector>
+
+#include "Core/Lua/LuaHelperClasses/LuaColonCheck.h"
+#include "Core/Lua/LuaVersionResolver.h"
+
+namespace Lua::LuaWheneverCallback
+{
+
+struct LuaThreadToFunctionReference
+{
+  lua_State* lua_thread;
+  int function_reference;
+};
+
+class LuaWhenever
+{
+public:
+  inline LuaWhenever() {}
+};
+
+static std::vector<LuaThreadToFunctionReference> list_of_functions_to_run;
+static bool list_of_functions_is_empty = true;
+static std::unique_ptr<LuaWhenever> instance = nullptr;
+static std::thread lua_whenever_callback_runner_thread;
+static std::mutex list_of_functions_lock;
+static std::condition_variable list_of_functions_condition_variable;
+static std::mt19937_64 rng_generator;
+static int temp_int;
+
+LuaWhenever* GetInstance()
+{
+  if (instance == nullptr)
+    instance = std::make_unique<LuaWhenever>(LuaWhenever());
+  return instance.get();
+}
+
+void InitLuaWheneverCallbackFunctions(lua_State* lua_state, const std::string& lua_api_version)
+{
+  rng_generator.seed(time(nullptr));
+  LuaWhenever** lua_whenever_instance_ptr_ptr =
+      (LuaWhenever**)lua_newuserdata(lua_state, sizeof(LuaWhenever*));
+  *lua_whenever_instance_ptr_ptr = GetInstance();
+  luaL_newmetatable(lua_state, "LuaWheneverFunctionsMetaTable");
+  lua_pushvalue(lua_state, -1);
+  lua_setfield(lua_state, -2, "__index");
+  std::array lua_whenever_functions_list_with_versions_attached = {
+
+      luaL_Reg_With_Version({"register", "1.0", Register}),
+      luaL_Reg_With_Version({"unregister", "1.0", Unregister})};
+
+  std::unordered_map<std::string, std::string> deprecated_functions_map;
+  AddLatestFunctionsForVersion(lua_whenever_functions_list_with_versions_attached,
+                               lua_api_version, deprecated_functions_map, lua_state);
+  lua_setglobal(lua_state, "Whenever");
+  lua_whenever_callback_runner_thread = std::thread(RunCallbacks);
+}
+
+int Register(lua_State* lua_state)
+{
+  list_of_functions_lock.lock();
+  list_of_functions_is_empty = false;
+  LuaColonOperatorTypeCheck(lua_state, "Whenever:register",
+                            "Whenever:register(functionName)");
+  lua_pushvalue(lua_state, 2);
+  int new_function_reference = luaL_ref(lua_state, LUA_REGISTRYINDEX);
+  lua_State* new_lua_thread = lua_newthread(lua_state);
```

Comment:
Do be aware that lua threads can't actually do multiprocessing (the name is a little misleading). Probally the cause of your issue. 

What you can do (and what we might eventually do for the gpu/cpu/gui threads) is use `lua_newstate` to create a completely independent state (with it's own set of globals) and then run that on a different OS thread.

However, that creates a huge nightmare of how to move data between different lua states, as lua has no built-in mechanism. Apparently [Lua Lanes](http://lualanes.github.io/lanes/) is a thing, no idea how good it is. 

The alternative option is to go the python "Global Interpreter Lock" route and have a global mutex for the entire lua state... I'm not a fan.

Reply:
> The alternative option is to go the python "Global Interpreter Lock" route and have a global mutex for the entire lua state... I'm not a fan.

I tried this approach also, but it still seems to cause a read-access exception. Whatever. I could always just get rid of the LuaWhenever option for now, and add it in to a later Lua release. That's when having a versioning system like this is convenient!

Reply:
What's the actual intention/usecase of LuaWhatever?

Reply:
@phire The point of LuaWhenever is to be able to accomplish your idea of running scripts that are independent of the thread that the OnFrameEnd script runs on.

Reply:
The amount of blocking this does for the UI and other threads also isn't great as well.

Reply:
Ah right. I probably did mention that somewhere along the way. 

In that case we can go with a reasonably simplistic API that meshes nicely with the fact that we need a fresh lua context. 

Lets steal some ideas and (and naming) from the web workers an worker threads in javascript and nodejs (the javascript ecosystem is a good place to look, has the same kind of limitation with wanting to avoid blocking)

Make it so the worker function has zero access to the main script's globals. Force the caller to pass everything it needs. The Lua Script host will copy the passed data into the worker thread's lua context. 

And then when the worker completes, it's return value is moved to back to the thread where it's needed.